### ISSUE TITLE
Single hyphen/dash argument for indicating stdin

### ISSUE THREAD

**Issue by moll** [pcapriotti/optparse-applicative#434]
Hey

Thanks for you continued maintenance of Optparse Applicative!

I'm having trouble finding a way to specify a single `-` as an expected value, a pattern used to refer to reading from standard input. I'd like the help text to show `(-|<path>)` and the parsing to result in either, let's say, `Nothing` given `-` on the command line and `Just path :: Maybe FilePath` if a path other than `-` was given. Is that possible at the level of Optparse?

In other words, I'm looking for something to replace `…` with:
```haskell
(Opt.flag' Nothing (…) <|>
(Just <$> Opt.argument Opt.str (Opt.metavar "<path>")))
```

This should then permit invoking an app with `./foo -` to read from stdin and `./foo file` to read from `file`, with the `--help` text nicely differentiating the two use-cases.

Thanks in advance!

**Comment by moll** [comment]
I'm thinking, if there were a mechanism for specifying expected constants, one could use that to both expect a single dash `-` argument and even support other advanced syntax in the style of `./foo rename <from> to <target>`.

I noticed `eitherReader`. Maybe that combined with `argument` and `metavar` could be used to fake what I'm going for...

**Comment by moll** [comment]
For now, I ended up going with:

```haskell
data PathArg path = StdinArg | PathArg path

Opt.argument
  (Opt.maybeReader parsePathArg)
  (mconcat [Opt.metavar "(-|<csv-path>)"])

parsePathArg "-" = Just StdinArg
parsePathArg path = Just (PathArg path)

```

**Comment by HuwCampbell** [comment]
That's more or less what I would have suggested.

**Comment by HuwCampbell** [comment]
There's a solution to this which you've figured out for yourself, but obviously this could be better documented.

I'll add a page to the wiki for it when I get some time.

**Comment by moll** [comment]
Thanks! And a built-in constant parser may be useful in the future, too, as my solution is rather verbose.
